### DISCUSSION TITLE
Idea: Input Field Hidden

### DISCUSSION THREAD

**Discussion by jonathan-martz** [pocketbase/pocketbase#1835] [Q&A]
Iam a custom type of users (db_user) and in my collection the field for password is show.
So i can see every password of every user, yeah currently no hash or something like that.

Would it make sense to make a Field Type like *hidden* where only ****** will be shown ?

**Comment by ganigeorgiev** [comment]
Without context I'm not sure that I understand the use case for it.

Passwords should be hashed and I can't think of scenario where storing them as plain text would be useful.
If you want a custom type of users, you can create a new auth collection (all auth collections come with system and hidden `passwordHash` field that stores the user's password as bcrypt hash) - https://pocketbase.io/docs/collections/#auth-collection.

Eventually we can add a `secret` type field to store some symmetric encrypted sensitive information but still I'm not sure whether its usage could be generalized. How this field will be managed? Will regular users, who have access to the collection record, be able to read the plain value? How the encryption key will be handled and who should be able to modify it?

Feel free to provide more details how you are planning to use the field, but for now there are too many uncertainties and it remains out of the scope.

**Reply by programonaut** [reply]
Hi,
I think it would be useful to have the option to convert a column into a secret type. Meaning I, as the admin, cannot see the data stored inside the field, and only the user is able to access it. For example, the user has to store some sort of key that the admin should not be able to see.

I tried replicating the passwordHash mechanism but sadly did not 100% understand how it works.

**Reply by JokerQyou** [reply]
This is useful for some write-only values. One possible use case is:

- Users can provide application keys and secrets to the server, so the server can do automation for them (or so that the server can integrate with external services).
- The server admin should not be able to see these values. The user (who filled these value into database) can see these values.
- Alternatively, there is such use case in which the user themselves can't see these values either. They can update them, but cannot read them out. The server admin can't view these values either. Only the server itself can read these values.

For example if users create personal GitHub token and submit them for use to fetch stats for their private repository, surely it's not safe if the server (written with PocketBase) admin can read these value, that would be a security issue. Users themselves may or may not be able to see their tokens, that's up to the server developer to decide.

**Reply by ganigeorgiev** [reply]
@JokerQyou Unfortunately this doesn't really clarify anything.

The statements "so that the server can integrate with external services" and "The server admin should not be able to see these values" contradict with each other.

Superuser have access to anything, including direct/raw access to the database (they can even load a malicous backup if they wanted to).

Again, for me it is still not clear how a "secret" field would be managed and used in a general way. If you disagree feel free to create a PoC with your suggestions but as mentioned in https://github.com/pocketbase/pocketbase/issues/2798 this for now remains out of the scope.

**Reply by JokerQyou** [reply]
> The statements "so that the server can integrate with external services" and "The server admin should not be able to see these values" contradict with each other.
> 
> Superuser have access to anything, including direct/raw access to the database (they can even load a malicous backup if they wanted to).

I was talking about using PocketBase as a framework (extending with Golang), so the server (the custom Golang code) can use user-supplied secrets to call external APIs, but I'd like to prevent the admin from seeing the secret values via the collections API or the management UI (the `/_/` dashboard).

Of course, if the admin have access to the database directly (access to the machine that runs PocketBase, or they could just dump a database backup), the secret field type alone would not be a viable solution. I realize this now.

For secret values I'd better store them in external services specifically designed to handle them, and link them with IDs in PocketBase.

**Comment by qgustavor** [comment]
For the case a non-password encrypted `secret` type field gets implemented, here's a blog post on database cryptography: https://soatok.blog/2023/03/01/database-cryptography-fur-the-rest-of-us/

The furry theming might look weird at first, but it's a good blog. It includes some pitfalls of simple solutions like the "We could use a tried-and-true encryption algorithm like AES-256" suggested in #2798, includes info about searching and querying encrypted data, and even includes a real world example from MongoCrypt, MongoDB’s encryption library.

I should add that not that AES should not be used, but AES as a starting point is bad, it's way too easy to misuse, the blog post includes examples. Well, crypto in general is hard: an example of mine is that I noticed and reported way too many files being encrypted with an all-zeroes AES key in mega.nz which turned to be because a popular badly implemented client. There are a lot of things to consider! I suggest you to consider using a well tested crypto library that already takes care of key generation (to avoid the issue from the mega client), nonces (nonce reuse is really bad) and contexts (to avoid a confused deputy attack). Maybe [Tink](https://developers.google.com/tink) or [Hydrogen](https://github.com/jedisct1/libhydrogen).

**Reply by ganigeorgiev** [reply]
@qgustavor It is not the encryption itself that prevents us to implement this type of field. It is the its actual use case that I'm not sure how things will work out (please see the questions in https://github.com/pocketbase/pocketbase/discussions/1835#discussioncomment-4946404). If you have a real use case and an idea for a workflow, please share it in https://github.com/pocketbase/pocketbase/issues/2798.

**Reply by qgustavor** [reply]
I understand, that's why _in the case it gets implemented_. For the application I'm working now I don't need it but I think someone will find a use case for that since there is a plugin for MongoDB to handle that. I'm more worried about security since it's something hard to get right and can be hard to fix once deployed. If that's not planned, I'm fine with that.

**Comment by JokerQyou** [comment]
This is actually possible with the new [`onRecordEnrich` hook](https://pocketbase.io/docs/go-event-hooks/#onrecordenrich). Basically this hook allows you to replace field values whenever they are fetched from database. If you replace certain fields with `***` using this hook, not event the superadmin can view its original value. The value is still stored correctly in the database, it's just transformed when querying with APIs. You can still fetch raw values (with custom code (using `App` instance) therefore bypassing this hook). I'm not sure if the backup file will contain the raw value or not, though.
